### DISCUSSION TITLE
Cache loader and automatic refresh

### DISCUSSION THREAD

**Discussion by mrodal** [infinispan/infinispan#10987] [Q&A] [answered]
Hi I wanted to know if there's a way to assign a function to the get(key) so that if the key is not cached, the function is executed, a read-through pattern. I know about NonBlockingStore but I dont know if its meant for this, all we need is to execute a function to get the value

And related to that, if there's a way to automatically refresh values? something like this: https://docs.oracle.com/cd/E15357_01/coh.360/e15723/cache_rtwtwbra.htm#COHDG198

Here's a simple example with Caffeine that provides this:
```java
LoadingCache<Key, Graph> graphs = Caffeine.newBuilder()
    .maximumSize(10_000)
    .refreshAfterWrite(1, TimeUnit.MINUTES) // Automatic refresh interval
    .build(key -> createExpensiveGraph(key)); // Key getter

```
https://github.com/ben-manes/caffeine/wiki/Refresh

Thanks

**Comment by wburns** [comment, marked as answer]
Currently, the best way to do this is via a custom NonBlockingStore implementation. There was interest in adding a method to the PersistenceConfigurationBuilder class to have a simple load method similar to Caffeine (that can be provided as a lambda), but this was never explored. Internally it would just use a NonBlockingStore implementation that only invokes the method.

Here is some example code to use for a store.

```java
public class GetStore implements NonBlockingStore<Object, Object> {
   @Override
   public Set<Characteristic> characteristics() {
      // TODO: optionally add SHAREABLE or SEGMENTABLE
      return EnumSet.of(Characteristic.READ_ONLY);
   }

   @Override
   public CompletionStage<Void> start(InitializationContext ctx) {
      return CompletableFutures.completedNull();
   }

   @Override
   public CompletionStage<Void> stop() {
      return CompletableFutures.completedNull();
   }

   @Override
   public CompletionStage<Void> write(int segment, MarshallableEntry<?, ?> entry) {
      throw new UnsupportedOperationException();
   }

   @Override
   public CompletionStage<Boolean> delete(int segment, Object key) {
      throw new UnsupportedOperationException();
   }

   @Override
   public CompletionStage<Void> clear() {
      throw new UnsupportedOperationException();
   }

   @Override
   public CompletionStage<MarshallableEntry<Object, Object>> load(int segment, Object key) {
      // TODO: implement here
      return null;
   }
}
```

The harder part is that all of our stores currently require a configuration builder but you should be able to use the CustomStoreConfigurationBuilder class to get around this.
